# Homework 3 - Optimization for Decision Making

This project solves a diet optimization problem using linear programming. The goal is to meet minimum daily vitamin and mineral requirements while minimizing total cost.

### Continous LP Model (Allows fractions)

In [ ]:
import pulp as pp
import pandas as pd


# Initializes the model
model = pp.LpProblem(name = 'vitamin-problem',
                     sense = pp.LpMinimize)

# Declares the variables

# SuperVit
sv = pp.LpVariable(name = 'super_vit',
                          lowBound = 0,
                          cat='integer')

# New Health
nh = pp.LpVariable(name = 'new_health',
                          lowBound = 0,
                          cat = 'integer')

# Create objective function to optimize
sv_coeff = .2
nh_coeff = .3
obj_func = sv_coeff * sv + nh_coeff * nh

# Assign Constraints

C1 = pp.LpConstraint(name = 'vitaminC',
                     e = 20 * sv + 30 * nh,
                     rhs = 60,
                     sense = pp.LpConstraintGE)

C2 = pp.LpConstraint(name = 'calcium',
                     e = 500 * sv + 250 * nh,
                     rhs = 1000,
                     sense=pp.LpConstraintGE)

C3 = pp.LpConstraint(name = 'iron',
                     e = 9 * sv + 2 * nh,
                     rhs = 18,
                     sense=pp.LpConstraintGE)

C4 = pp.LpConstraint(name = 'niacin',
                     e = 2 * sv + 10 * nh,
                     rhs = 20,
                     sense=pp.LpConstraintGE)

C5 = pp.LpConstraint(name = 'magnesium',
                     e = 60 * sv + 90 * nh,
                     rhs = 360,
                     sense=pp.LpConstraintGE)

# Build the model

model += obj_func
model += C1
model += C2
model += C3
model += C4
model += C5

# Solve the model

status = model.solve()

# List LP continous results

Results={"Optimal Solution to minimize cost":pp.value(model.objective)}
Results.update({v.name: v.varValue for v in model.variables()})
Results

pd.DataFrame.from_dict(Results,
                       orient = 'index',
                       columns = ['info']).map('{:,.2f}'.format)




info
Optimal Solution to minimize cost  1.20
new_health                         3.13
super_vit                          1.30

### Whole Number (Integer) LP Results

In [ ]:
import pulp as pp

# Builds model
model = pp.LpProblem("diet_problem", pp.LpMinimize)

# Assigns variables
sv = pp.LpVariable("super_vit", lowBound=0, cat=pp.LpInteger)
nh = pp.LpVariable("new_health", lowBound=0, cat=pp.LpInteger)

# Builds model
model += 0.2*sv + 0.3*nh

model += 20*sv + 30*nh >= 60, "vitaminC"
model += 500*sv + 250*nh >= 1000, "calcium"
model += 9*sv + 2*nh >= 18, "iron"
model += 2*sv + 10*nh >= 20, "niacin"
model += 60*sv + 90*nh >= 360, "magnesium"

# Solves model
status = model.solve(pp.PULP_CBC_CMD(msg=True))

# Lists Integer LP model's results
print("Status:", pp.LpStatus[status])

for v in model.variables():
    print(v.name, v.cat, v.value())

print("Cost:", pp.value(model.objective))

Status: Optimal
new_health Integer 2.0
super_vit Integer 3.0
Cost: 1.2000000000000002


***
### How many of each vitamin to meet the minimum requirements while spending the least amount of money?

I’m reporting both the continuous LP solution (fractional tablets allowed) and the integer solution (whole tablets only).

* **Continuous LP (fractional tablets allowed)**
The optimal solution is SuperVit = 1.304 and NewHealth = 3.130, for a minimum daily cost of $1.20.

* **Integer solution (whole tablets only)**
If tablets must be whole numbers, an optimal solution is SuperVit = 3 and NewHealth = 2, with the same minimum daily cost of $1.20.

Both solutions achieve the same minimum cost, but the continuous model allows fractional quantities while the integer model restricts purchases to whole tablets.